scene/3d/physics_body.cpp

```diff
@@ -1150,10 +1150,20 @@ Vector3 KinematicBody::move_and_slide(const Vector3 &p_linear_velocity, const Ve
 		}
 	}
 
+	Vector3 floor_motion = floor_velocity;
+	if (on_floor && on_floor_body.is_valid()) {
+		//this approach makes sure there is less delay between the actual body velocity and the one we saved
+		PhysicsDirectBodyState *bs = PhysicsServer::get_singleton()->body_get_direct_state(on_floor_body);
+		if (bs) {
+			floor_motion = bs->get_linear_velocity();
+		}
+	}
+
 	// Hack in order to work with calling from _process as well as from _physics_process; calling from thread is risky
-	Vector3 motion = (floor_velocity + lv) * (Engine::get_singleton()->is_in_physics_frame() ? get_physics_process_delta_time() : get_process_delta_time());
+	Vector3 motion = (floor_motion + lv) * (Engine::get_singleton()->is_in_physics_frame() ? get_physics_process_delta_time() : get_process_delta_time());
```

Comment:
Here motion refers to a lengths (velocity x time), so using `floor_motion` for this velocity local is confusing.

Reply:
It's the same name that's used in the 2D version. I'll rename them both to `current_floor_velocity`.
I'll also use the opportunity to rename the `lv` and `lv_n` variables to `body_velocity` and `body_velocity_normal`